File: leetcode3/이진희/3707_Equal Score Substrings.java
/*

1. 아이디어 : 미리 모든 알파벳의 합을 구한다. 이후, 슬라이딩 윈도우 기법으로 직접 잘라보며 값이 같아지는지 판단한다.

2. 시간복잡도 : O(2*N)

3. 자료구조/알고리즘 : 슬라이딩 윈도우

 */

class Solution {
    public boolean scoreBalance(String s) {
        // 잘라서 합이 같아야함
        // 1부터 n-1까지 자르기 가능

        int leftSum = 0;
        int rightSum = 0;
        for(int i=0; i<s.length(); i++) {
            leftSum+=s.charAt(i)-'a'+1;
        }

        for(int i=0; i<s.length(); i++) {
            leftSum-=s.charAt(i)-'a'+1;
            rightSum+=s.charAt(i)-'a'+1;

            if(leftSum == rightSum) return true;
        }

        return false;
    }
}

File: leetcode3/최원준/3707. Equal Score Substrings.py
#

'''
1. 아이디어 :
투포인터를 사용한다.
두 포인터가 겹칠때까지 움직이며 값을 계산한다.

2. 시간복잡도 :
    O(n)

3. 자료구조/알고리즘 :
투포인터

'''

class Solution:
    def scoreBalance(self, s: str) -> bool:
        n = len(s)
        left, right = 0, n-1
        l_val = 0
        r_val = 0

        while left <= right:
            if l_val <= r_val:
                l_val+=ord(s[left])-96
                left+=1
            else:
                r_val+=ord(s[right])-96
                right-=1
        return l_val == r_val
        
            
